Source/DafnyRuntime/DafnyRuntimeRust/src/lib.rs

```diff
@@ -1650,24 +1650,47 @@ impl<V: DafnyTypeEq> PartialOrd<Multiset<V>> for Multiset<V> {
     fn partial_cmp(&self, other: &Multiset<V>) -> Option<Ordering> {
         match self.cardinality().cmp(&other.cardinality()) {
             Ordering::Less => {
-                for value in other.data.keys() {
-                    if !self.contains(value) || self.get(value) > other.get(value) {
+                for value in self.data.keys() {
+                    if !other.contains(value) || self.get(value) > other.get(value) {
                         return None;
                     }
                 }
                 Some(Ordering::Less)
             }
             Ordering::Equal => {
+                let mut has_less = false;
+                let mut has_greater = false;
                 for value in self.data.keys() {
-                    if self.get(value) != other.get(value) {
+                    if !other.contains(value) {
                         return None;
                     }
+                    let self_count = self.get(value);
+                    let other_count = other.get(value);
+                    if self_count < other_count {
+                        if has_greater {
+                            return None;
+                        }
+                        has_less = true;
+                    } else if self_count > other_count {
+                        if has_less {
+                            return None;
+                        }
+                        has_greater = true;
+                    }
+                }
+                if has_less && has_greater {
+                    None // should be unreachable
+                } else if has_less {
+                    Some(Ordering::Less)
```

Comment:
Is it possible for a multiset X of the same cardinality as a multiset Y to be less in the ordering?

Consider this example
```
self == multiset{1, 1, 2}
other == multiset{1, 2, 3}
```
Your algorithm would return `Some(Ordering::Greater)` simply because `self` has more 1s than `other`, not accounting for `other` to contain 3.